libvast/builtins/operators/replace.cpp

```diff
@@ -51,10 +53,24 @@ struct bound_configuration {
   make(const type& schema, const configuration& config) {
     auto result = bound_configuration{};
     const auto& schema_rt = caf::get<record_type>(schema);
-    for (const auto& [extractor, value] : config.extractor_to_value)
+    for (const auto& [extractor, value] : config.extractor_to_value) {
+      // The config parsing never produces all possible alternatives of the data
+      // variant, e.g., addresses will be represented as strings. Because of
+      // that we need to re-parse the data if it's a string.
+      auto reparsed_value = [](auto value) {
+        const auto* str = caf::get_if<std::string>(&value);
+        if (!str)
+          return value;
+        auto result = to<data>(*str);
+        if (!result)
+          return value;
+        return std::move(*result);
+      }(value);
```

Comment:
Should this be a friend function of `data` more generally, maybe along the lines of `parse_if_string`?

Reply:
Why friend function and not just a normal free function? I would argue that the interface of `data` should be minimal and that parsing/printing is an orthogonal concern.

Reply:
Two, yes:
- Make it easy for the function to be discovered (either by looking at the header or at tooling), and
- Minimally improve compile times.

It's just generally a good practice. Free functions are less discoverable, so they tend to be overlooked if you don't know they exist. As the team grows we need to make sure that we have functions in sane places, as we cannot expect everyone to always read through all changes.

I know there are two camps for this discussion and that you specifically are very much in the other one here. :)

Reply:
Ultimately, I'd also be happy with just leaving the lambda here as-is for now.

Reply:
Given that we need to (slightly) rewrite these two operators anyways with the executor refactoring, we can use that opportunity for cleanup. I'll just rebase this now and leave the lambdas as they are.